### PR TITLE
GH-32206: [C++] GcsFileSystem::Make should return Result

### PR DESCRIPTION
**Pull request by trifleneurotic**
### What changes are included in this PR?

GcsFileSystem::Make now returns Result, with corresponding header & test changes.

### Are these changes tested?

Yes, `arrow-gcsfs-test` passed.

### Are there any user-facing changes?

Yes.

**This PR includes breaking changes to public APIs.**

* GitHub Issue: #32206

### CODE REVIEW

**Comment by trifleneurotic** [comment]
GH-32206: Added back a line I had erroneously removed (`port_` argument for `storage-testbench`).

File: cpp/src/arrow/filesystem/gcsfs.cc [hunk 1]
```diff
@@ -962,7 +962,7 @@ Result<std::shared_ptr<io::OutputStream>> GcsFileSystem::OpenAppendStream(
   return Status::NotImplemented("Append is not supported in GCS");
 }
 
-std::shared_ptr<GcsFileSystem> GcsFileSystem::Make(const GcsOptions& options,
+Result<std::shared_ptr<GcsFileSystem>> GcsFileSystem::Make(const GcsOptions& options,
```
Comment (mapleFU) [inline cpp/src/arrow/filesystem/gcsfs.cc:965]: So this is for compatible reason? 🤔

Comment (trifleneurotic) [reply cpp/src/arrow/filesystem/gcsfs.cc:965]: Yes correct, as per the original issue: "For consistency with S3FileSystem" which, incidentally, makes it consistent with other similar filesystems.

File: cpp/src/arrow/filesystem/gcsfs.h [hunk 2]
```diff
@@ -232,7 +232,7 @@ class ARROW_EXPORT GcsFileSystem : public FileSystem {
 
   /// Create a GcsFileSystem instance from the given options.
   // TODO(ARROW-16884): make this return Result for consistency
```
Comment (kou) [inline cpp/src/arrow/filesystem/gcsfs.h:234]: Could you remove this comment?

**Review by kou — COMMENTED** [review]
`gcsfs.h` is a public header. So this is a breaking change. Could you keep `**This PR includes breaking changes to public APIs.**` in the pull request template?

File: cpp/src/arrow/filesystem/gcsfs_test.cc [hunk 3]
```diff
@@ -271,7 +271,7 @@ class TestGCSFSGeneric : public GcsIntegrationTest, public GenericFileSystemTest
   void SetUp() override {
     ASSERT_NO_FATAL_FAILURE(GcsIntegrationTest::SetUp());
     auto bucket_name = RandomBucketName();
-    gcs_fs_ = GcsFileSystem::Make(TestGcsOptions());
+    ASSERT_OK_AND_ASSIGN(std::shared_ptr<GcsFileSystem> gcs_fs_, GcsFileSystem::Make(TestGcsOptions()));
```
Comment (kou) [inline cpp/src/arrow/filesystem/gcsfs_test.cc:274]:
```suggestion
    ASSERT_OK_AND_ASSIGN(gcs_fs_, GcsFileSystem::Make(TestGcsOptions()));
```

File: cpp/src/arrow/filesystem/gcsfs_test.cc [hunk 4]
```diff
@@ -489,17 +489,13 @@ TEST(GcsFileSystem, FileSystemCompare) {
   a_options.project_id = "test-only-invalid-project-id";
   auto a = GcsFileSystem::Make(a_options);
   EXPECT_THAT(a, NotNull());
-  EXPECT_TRUE(a->Equals(*a));
```
Comment (kou) [inline cpp/src/arrow/filesystem/gcsfs_test.cc:492]: Why do we need to remove this?

Should we change `auto a = GcsFileSystem::Make(a_options);` to `ASSERT_OK_AND_ASSIGN(auto a, GcsFileSystem::Make(a_options);` instead?

File: cpp/src/arrow/filesystem/gcsfs_test.cc [hunk 5]
```diff
@@ -622,7 +618,7 @@ TEST(GcsFileSystem, ObjectMetadataRoundtrip) {
 }
 
 TEST_F(GcsIntegrationTest, GetFileInfoBucket) {
-  auto fs = GcsFileSystem::Make(TestGcsOptions());
+  ASSERT_OK_AND_ASSIGN(std::shared_ptr<GcsFileSystem> fs, GcsFileSystem::Make(TestGcsOptions()));
```
Comment (kou) [inline cpp/src/arrow/filesystem/gcsfs_test.cc:621]: Could you keep using `auto`?

**Comment by trifleneurotic** [comment]
> `gcsfs.h` is a public header. So this is a breaking change. Could you keep `**This PR includes breaking changes to public APIs.**` in the pull request template?

This has been done.

File: cpp/src/arrow/filesystem/gcsfs.h (hunk 2 above)
Comment (trifleneurotic) [reply cpp/src/arrow/filesystem/gcsfs.h]: Complete! Ref 4d6687ad52ac1734fab38e092ebc149fe87b09ce.

File: cpp/src/arrow/filesystem/gcsfs_test.cc (hunk 3 above)
Comment (trifleneurotic) [reply cpp/src/arrow/filesystem/gcsfs_test.cc]: Complete! Ref 4d6687ad52ac1734fab38e092ebc149fe87b09ce.

File: cpp/src/arrow/filesystem/gcsfs_test.cc (hunk 4 above)
Comment (trifleneurotic) [reply cpp/src/arrow/filesystem/gcsfs_test.cc:492]: There is no need to remove this with your suggested change. That change has been made, and the above line has been restored. Ref 4d6687ad52ac1734fab38e092ebc149fe87b09ce.

**Comment by trifleneurotic** [comment]
Tests were re-run locally as a result of the local compilation of commit 4d6687ad52ac1734fab38e092ebc149fe87b09ce; `arrow-gcsfs-test` was successful.

File: cpp/src/arrow/filesystem/gcsfs.h [hunk 6]
```diff
@@ -230,9 +230,7 @@ class ARROW_EXPORT GcsFileSystem : public FileSystem {
       const std::string& path,
       const std::shared_ptr<const KeyValueMetadata>& metadata) override;
 
-  /// Create a GcsFileSystem instance from the given options.
```
Comment (kou) [inline cpp/src/arrow/filesystem/gcsfs.h:233]: We don't need to remove this comment too. Could you revert this change?

File: cpp/src/arrow/filesystem/gcsfs_test.cc [hunk 7]
```diff
@@ -271,7 +271,8 @@ class TestGCSFSGeneric : public GcsIntegrationTest, public GenericFileSystemTest
   void SetUp() override {
     ASSERT_NO_FATAL_FAILURE(GcsIntegrationTest::SetUp());
     auto bucket_name = RandomBucketName();
-    gcs_fs_ = GcsFileSystem::Make(TestGcsOptions());
+    ASSERT_OK_AND_ASSIGN(std::shared_ptr<GcsFileSystem> gcs_fs_,
```
Comment (kou) [inline cpp/src/arrow/filesystem/gcsfs_test.cc:274]:
```suggestion
    ASSERT_OK_AND_ASSIGN(gcs_fs_,
```

File: cpp/src/arrow/filesystem/gcsfs_test.cc [hunk 8]
```diff
@@ -497,9 +498,6 @@ TEST(GcsFileSystem, FileSystemCompare) {
   b_options.project_id = "test-only-invalid-project-id";
   auto b = GcsFileSystem::Make(b_options);
   EXPECT_THAT(b, NotNull());
-  EXPECT_TRUE(b->Equals(*b));
-
-  EXPECT_FALSE(a->Equals(*b));
```
Comment (kou) [inline cpp/src/arrow/filesystem/gcsfs_test.cc:502]: Can we revert these changes?

File: cpp/src/arrow/filesystem/gcsfs_test.cc [hunk 9]
```diff
@@ -632,7 +630,7 @@ TEST_F(GcsIntegrationTest, GetFileInfoBucket) {
 TEST_F(GcsIntegrationTest, GetFileInfoObjectWithNestedStructure) {
   // Adds detailed tests to handle cases of different edge cases
   // with directory naming conventions (e.g. with and without slashes).
-  auto fs = GcsFileSystem::Make(TestGcsOptions());
+  ASSERT_OK_AND_ASSIGN(std::shared_ptr<GcsFileSystem> fs, GcsFileSystem::Make(TestGcsOptions()));
```
Comment (kou) [inline cpp/src/arrow/filesystem/gcsfs_test.cc:633]: Can we keep `auto`?

BTW, we have many similar changes. Could you do self-review again?

File: cpp/src/arrow/filesystem/gcsfs.h (hunk 6 above)
Comment (trifleneurotic) [reply cpp/src/arrow/filesystem/gcsfs.h:233]: Reverted! Ref e5b606e462cb362ef49092eb10797d166e2e60fd.

File: cpp/src/arrow/filesystem/gcsfs_test.cc (hunk 7 above)
Comment (trifleneurotic) [reply cpp/src/arrow/filesystem/gcsfs_test.cc]: Modified! e5b606e462cb362ef49092eb10797d166e2e60fd.

File: cpp/src/arrow/filesystem/gcsfs_test.cc (hunk 8 above)
Comment (trifleneurotic) [reply cpp/src/arrow/filesystem/gcsfs_test.cc]: Reverted! Ref e5b606e462cb362ef49092eb10797d166e2e60fd.

File: cpp/src/arrow/filesystem/gcsfs_test.cc (hunk 9 above)
Comment (trifleneurotic) [reply cpp/src/arrow/filesystem/gcsfs_test.cc]: Made changes on all appropriate lines. Ref e5b606e462cb362ef49092eb10797d166e2e60fd.

**Comment by trifleneurotic** [comment]
Tests were again re-run locally as a result of the local compilation of commit e5b606e462cb362ef49092eb10797d166e2e60fd; `arrow-gcsfs-test` was successful.

**Comment by kou** [comment]
Could you fix lint failures by `pre-commit run --show-diff-on-failure --color=always --all`?

**Comment by trifleneurotic** [comment]
> Could you fix lint failures by `pre-commit run --show-diff-on-failure --color=always --all`?

I think those failures should be fixed now (ref 07c36f5786888c1e2398cda5911bc4f986fe30f4 & bc88c9ffe39d1d8c5e3ebb8b5ecb63b7aedf975e).

**Comment by trifleneurotic** [comment]
I apologize sincerely for the linting back-and-forth; just committed another change (ref f0ba9c5cec42962c5b9fde69d3734ec6367ce9ef) to match method signature style to nearby method. I appreciate everyone's guidance & advice as I work towards an acceptable PR.

**Comment by kou** [comment]
Why did you fix it manually?
Did `pre-commit run --show-diff-on-failure --color=always --all` not work well?

**Comment by trifleneurotic** [comment]
> Why did you fix it manually? Did `pre-commit run --show-diff-on-failure --color=always --all` not work well?

That's correct it did not, but after further investigation it turns out it was a local issue on my part - my apologies. That issue has been resolved & I was able to correctly capture & push `pre-commit`'s changes (ref 2090bbba59471f4a1eba426557bc90674a88b308).

**Comment by kou** [comment]
It's related:

https://github.com/apache/arrow/actions/runs/11499533014/job/32034937833?pr=44503#step:7:6721

```text
g++ -std=gnu++17 -I"/usr/share/R/include" -DNDEBUG -I/usr/local/include    -DARROW_R_WITH_PARQUET -DARROW_R_WITH_DATASET -DARROW_R_WITH_ACERO -DARROW_R_WITH_SUBSTRAIT -DARROW_R_WITH_JSON -DARROW_R_WITH_S3 -DARROW_R_WITH_GCS -Werror -I'/usr/local/lib/R/site-library/cpp11/include'     -fpic  -g -O2 -fno-omit-frame-pointer -mno-omit-leaf-frame-pointer -ffile-prefix-map=/build/r-base-YnquTi/r-base-4.4.1=. -fstack-protector-strong -fstack-clash-protection -Wformat -Werror=format-security -fcf-protection -fdebug-prefix-map=/build/r-base-YnquTi/r-base-4.4.1=/usr/src/r-base-4.4.1-3.2404.0 -Wdate-time -D_FORTIFY_SOURCE=3   -c type_infer.cpp -o type_infer.o
filesystem.cpp: In function ‘std::shared_ptr<arrow::fs::GcsFileSystem> fs___GcsFileSystem__Make(bool, cpp11::list)’:
filesystem.cpp:433:33: error: could not convert ‘arrow::fs::GcsFileSystem::Make(const arrow::fs::GcsOptions&, const arrow::io::IOContext&)(io_context)’ from ‘arrow::Result<std::shared_ptr<arrow::fs::GcsFileSystem> >’ to ‘std::shared_ptr<arrow::fs::GcsFileSystem>’
  433 |   return fs::GcsFileSystem::Make(gcs_opts, io_context);
      |          ~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~
      |                                 |
      |                                 arrow::Result<std::shared_ptr<arrow::fs::GcsFileSystem> >
make: *** [/usr/lib/R/etc/Makeconf:204: filesystem.o] Error 1
```

Hmm. It may be better that we keep the old signature with deprecation for easy to migration. Something like:

```diff
diff --git a/cpp/src/arrow/filesystem/gcsfs.h b/cpp/src/arrow/filesystem/gcsfs.h
index f1fbc95bf9..3789da18ec 100644
--- a/cpp/src/arrow/filesystem/gcsfs.h
+++ b/cpp/src/arrow/filesystem/gcsfs.h
@@ -231,10 +231,14 @@ class ARROW_EXPORT GcsFileSystem : public FileSystem {
       const std::shared_ptr<const KeyValueMetadata>& metadata) override;
 
   /// Create a GcsFileSystem instance from the given options.
-  // TODO(ARROW-16884): make this return Result for consistency
+  ARROW_DEPRECATED("Deprecated in 19.0.0. Use the Result version instead.")
   static std::shared_ptr<GcsFileSystem> Make(
       const GcsOptions& options, const io::IOContext& = io::default_io_context());
 
+  /// Create a GcsFileSystem instance from the given options.
+  static Result<std::shared_ptr<GcsFileSystem>> Make(
+      const GcsOptions& options, const io::IOContext& = io::default_io_context());
+
  private:
   explicit GcsFileSystem(const GcsOptions& options, const io::IOContext& io_context);
 ```

FYI: You can check CI results on your for by enabling GitHub Actions on your fork.

**Comment by trifleneurotic** [comment]
Thank you all for your advice & time so far in reviewing my PR.

Just to give a quick update:

For my latest commit (ref 83a49be838c2ffb8de8dc4e6574e9523535a70e4), I have done the following:

- local build with the `ninja-debug-filesystems` preset
- ran all `cmake` tests (all successful)
- ran `pre-commit` (all green)

Of course, after that, I did the requisite `pull`s & `rebase` before `push`. This was an older issue so, even though it was a `good-first-issue` at the time, there may be more tendrils now connecting to other places in the codebase. I've been able to reduce the number of post-`push` job failures but haven't been able to eliminate them yet.

Thoughts on moving forward?

**Comment by trifleneurotic** [comment]
I am currently readying a commit/push to address the issues with R. Please stand by, but of course, feel free to comment in the meantime if desired.

**Comment by trifleneurotic** [comment]
Hello! And thank you all for your advice & time so far in reviewing my PR.

Just to give a quick update:

For my latest commit (ref 6e9979df8fe301579c6825ae75bdfa049a6f6259), I have done the following:

- local build with the `ninja-debug-filesystems` preset
- ran all R `testthat` tests (`gcs` successful)
- ran all `cmake` tests (`arrow-gcsfs-test` successful)
- ran `pre-commit` (changes made)

Of course, after that, I did the requisite `pull`s & `rebase` before `push`. This was an older issue so, even though it was a `good-first-issue` at that time, there may be more tendrils now connecting to other places in the codebase. I've been able to reduce the number of post-`push` job failures on my fork but haven't been able to eliminate them yet.

Thoughts on moving forward?

File: r/src/filesystem.cpp [hunk 10]
```diff
@@ -429,8 +429,8 @@ std::shared_ptr<fs::GcsFileSystem> fs___GcsFileSystem__Make(bool anonymous,
   }
 
   auto io_context = MainRThread::GetInstance().CancellableIOContext();
-  // TODO(ARROW-16884): update when this returns Result
-  return fs::GcsFileSystem::Make(gcs_opts, io_context);
+
+  return ValueOrStop(fs::GcsFileSystem::Make(gcs_opts, io_context));
```
Comment (kou) [inline r/src/filesystem.cpp:433]: The R bindings supports both of old Arrow C++ and new Arrow C++:

```suggestion
#if ARROW_VERSION_MAJOR >= 19
  return ValueOrStop(fs::GcsFileSystem::Make(gcs_opts, io_context));
#else
  return fs::GcsFileSystem::Make(gcs_opts, io_context);
#endif
```

File: cpp/src/arrow/filesystem/gcsfs.cc (hunk 1 above)
Comment (trifleneurotic) [reply cpp/src/arrow/filesystem/gcsfs.cc]: Done! Ref b27686b4ccfb49171ca52596ab28734024696dcf (although I had to use 18 as major version number for workflow to finish).

File: r/src/filesystem.cpp (hunk 10 above)
Comment (trifleneurotic) [reply r/src/filesystem.cpp]: Done! Ref b27686b4ccfb49171ca52596ab28734024696dcf (although I had to lower the major version to 18 to get workflow to finish).

File: r/src/filesystem.cpp [hunk 11]
```diff
@@ -429,8 +429,12 @@ std::shared_ptr<fs::GcsFileSystem> fs___GcsFileSystem__Make(bool anonymous,
   }
 
   auto io_context = MainRThread::GetInstance().CancellableIOContext();
-  // TODO(ARROW-16884): update when this returns Result
+
```
Comment (kou) [inline r/src/filesystem.cpp:432]: Could you remove needless diff?

```suggestion
```

File: r/src/filesystem.cpp (hunk 10 above)
Comment (kou) [reply r/src/filesystem.cpp]: OK. Then let's wait for 18.0.0 release. We will bump development version to 19 in the 18.0.0 release process.
FYI: You can monitor https://lists.apache.org/thread/c3t7wn32pp506925fm1bg15lrrk4jrz6 thread for the 18.0.0 release.

Comment (trifleneurotic) [reply r/src/filesystem.cpp]: Will monitor! Also removed unnecessary diff (ref 1d6b53ae38bfce51d828a0a473601e538dd1a4b0).

**Comment by trifleneurotic** [comment]
@kou 

Hello! I see that 18.0.0 showed up on the [Releases](https://github.com/apache/arrow/releases/tag/apache-arrow-18.0.0) page for Arrow. What would be next steps?

**Comment by kou** [comment]
Could you rebase on main?

**Comment by trifleneurotic** [comment]
> Could you rebase on main?

@kou

Complete! Ref 51cbac705f3b398e3b67e6931bfe9c0ebab39ead.

**Comment by trifleneurotic** [comment]
@kou 

Thank you for yoiur continued assistance.

I had two workflow job failures over on my fork (`NodeJS` & `Archery/Crossbow`), although here there was just one: `Python / AMD64 Conda Python 3.11 Pandas latest`. Would you happen to have any context or background as to these workflow job failures?

**Review by kou — APPROVED** [review]
+1

Thanks.
